### PR TITLE
Ability to hide patches of deleted files, significantly speeding up `magit-status` (#2982)

### PR DESCRIPTION
**Pull request by bramadams**
Hi,

Before this patch, simply updating a package like org-plus-contrib, then running `magit-status` on my .emacs.d git repo would take 20 seconds, followed by similar delays upon (un)staging of files due to refreshing the magit buffer. The main reason was that deleted files have their entire contents shown in the patch subsections of the (un)staged sections of `magit-status`. This issue is similar to for example https://github.com/magit/magit/issues/3808, and is also related to https://github.com/magit/magit/issues/2982.

With this patch, `magit-status` and subsequent refreshes only take a couple of seconds.

The only thing the patch is doing is to invoke "git diff" with the "-D" (--irreversible-delete) option for the (un)staged file sections, see https://git-scm.com/docs/diff-options/2.14.2#diff-options---irreversible-delete. This option still shows the deleted file in the summary of the diff, but excludes the actual contents of the file. As such the buffer of `magit-status` is much smaller, which speeds up `magit-status-refresh-buffer`, while one can still (un)stage deleted files as before. The only thing that has become impossible is opening the deleted file by hitting enter.

Since some people might not like this idea, the patch also introduces a defcustom named `magit-diff-hide-deleted-file-patches` to toggle whether or not deleted files should have their patch shown by `magit-status` for the (un)staged change sections. Default value is `nil`, i.e., the current (slow) behaviour.

Unfortunately, "git diff" does not have an equivalent argument for new files, hence large added files could still slow down `magit-status-refresh-buffer`.

In practice, though, this might not be as bad as it sounds: in case of renamed files
or, for example, updated MELPA packages, if one first stages the old, deleted version of a file or folder (which prompts the now faster refresh), then the new (untracked) version, git's rename detection will kick in, which automatically hides the contents of the renamed files and hence significantly speeds up the second refresh.

TL;DR Significantly faster magit refresh by hiding deleted files' patches during (un)staging; in case of renaming, first stage the deleted file/folder.

### CODE REVIEW

**Comment by tarsius** [comment]
I don't think a new Custom option is necessary. We can just add this argument to the two relevant transients.

You will have to reveal, enable and save the argument in both transients (on `D` and `d`).

**Comment by tarsius** [comment]
Certain transient popups hide some of their suffix commands by default. These additional suffixes can be [enabled interactively](https://magit.vc/manual/transient/Enabling-and-Disabling-Suffixes.html). TL;DR Enter the transient as usual and type <kbd>C-x l</kbd>.

<sub>This is a [canned response](https://help.github.com/en/github/writing-on-github/working-with-saved-replies); it may not apply 100% in this case.</sub>

**Comment by tarsius** [comment]
The infix arguments used by a certain transient popups can be [set/saved](https://magit.vc/manual/transient/Saving-Values.html) for future invocations.  TL;DR Enter the transient as usual, enable and/or disable arguments, and finally type <kbd>C-x s</kbd> or <kbd>C-x C-s</kbd>.

<sub>This is a [canned response](https://help.github.com/en/github/writing-on-github/working-with-saved-replies); it may not apply 100% in this case.</sub>

**Comment by bramadams** [comment]
Hi,

Works great, thanks for the improved fix!

BA

PS: Perhaps it's worth adding this functionality (+ the "stage deleted first" trick) to https://magit.vc/manual/magit/Performance.html, since it makes a big difference.

**Comment by tarsius** [comment]
Please feel free to suggest a paragraph. For me files being deleted isn't a regular experience, so it doesn't matter much for me. Which reminds me, I wanted to ask:

> Before this patch, simply updating a package like org-plus-contrib,

Why? Why would "updating" (`git pull`?) cause files to disappear from just the working tree but not `HEAD` (or the other way around I suppose)?

**Comment by bramadams** [comment]
> > Before this patch, simply updating a package like org-plus-contrib,
> 
> Why? Why would "updating" (`git pull`?) cause files to disappear from just the working tree but not `HEAD` (or the other way around I suppose)?

Sorry for the confusion: I meant "updating" in the emacs package manager sense, which boils down to deleting the old version of, say org-plus-contrib, and installing the new version. Since I version-control the installed packages in my emacs config, to commit the package update the Magit buffer for my repo shows the deleted version in the unstaged section (by default with full contents shown, hence slow) and the new version in the untracked section (just folder name, hence quick). With the -D tweak, the Magit buffer now refreshes almost instantly for this use case.

> Please feel free to suggest a paragraph. For me files being deleted isn't a regular experience, so it doesn't matter much for me. 

OK, here is the suggested text to replace the start of the 3rd paragraph of 9.2.2 on https://magit.vc/manual/magit/Performance.html#Performance:
***
<p>Meanwhile, there are a number of things that can be done to speed up the Magit buffer refresh. If you often deal with deleted files (folders), or with renamed files (folders), you can disable the preview of the contents of the deleted/renamed files (folders) by enabling the <code>--irreversible-delete</code> option of <code>git diff</code> (possible from version 3.0 of Magit on). The only downside of this performance optimization is that hitting enter on (un)staged deleted/renamed file will no longer open that file.<p>

<p>To enable <code>--irreversible-delete</code>, open the Magit status buffer, then open the <code>magit-diff</code> transient using <code>d</code>. In this transient, type <code>C-x l</code>, followed by <code>-D</code>, then input <code>1</code>, hit enter, then type again <code>-D</code>. Now you just need to make this option permanent for future sessions by typing <code>C-x C-s</code>. If desired, you can hide the (enabled) option again using <code>C-x l</code>, followed by <code>-D</code>, then input <code>5</code>, hit enter).</p>

<p>Unfortunately, there is no equivalent option for newly added files. However, in case of renamed files (folders; typical example is when committing an upgraded emacs package in your personal configuration repo), you can still speed up the Magit buffer refresh by making a habit of first staging the deleted file (folder) before staging the new file (folder). Git's internal rename detection will automatically hide the contents of the renamed file (folder), making the Magit buffer refresh much snappier.</p>

<p>If this does not work well in your case, a more extreme performance optimization is to only automatically refresh the current Magit buffer
